fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LucenOnlineIndexingTest.java

```diff
@@ -937,5 +939,70 @@ void testRecordUpdateReducedMergeForcingAgileTimeQuota() {
         assertTrue(newLength < oldLength);
         assertTrue(timeCommitCount > 0);
     }
+
+    private static class TerribleRebanalceIndexMaintainer extends LuceneIndexMaintainer {
+        private final IndexMaintainerState state;
+        private static int pseodoFound = 0;
+
+        protected TerribleRebanalceIndexMaintainer(final IndexMaintainerState state) {
+            super(state, state.context.getExecutor());
+            this.state = state;
+        }
+
+        @Override
+        public CompletableFuture<Void> mergeIndex() {
+            final IndexDeferredMaintenanceControl mergeControl = state.store.getIndexDeferredMaintenanceControl();
+            mergeControl.setLastStep(IndexDeferredMaintenanceControl.LastStep.REBALANCE);
+            final int documentCount = mergeControl.getRepartitionDocumentCount();
+            if (documentCount <= 0) {
+                mergeControl.setRepartitionDocumentCount(16);
+                throw new FDBException("transaction_too_old", FDBError.TRANSACTION_TOO_OLD.code());
+            }
+            if (documentCount > 2) {
+                throw new FDBException("transaction_too_old", FDBError.TRANSACTION_TOO_OLD.code());
+            }
+            return AsyncUtil.DONE;
+        }
+    }
+
+    @AutoService(IndexMaintainerFactory.class)
+    public static class TerribleRebanalceIndexMaintainerFactory implements IndexMaintainerFactory {
+        @Nonnull
+        @Override
+        public Iterable<String> getIndexTypes() {
+            return Collections.singletonList("terribleRebalance");
+        }
+
+        @Nonnull
+        @Override
+        public IndexValidator getIndexValidator(Index index) {
+            return new IndexValidator(index);
+        }
+
+        @Nonnull
+        @Override
+        public IndexMaintainer getIndexMaintainer(@Nonnull IndexMaintainerState state) {
+            return new TerribleRebanalceIndexMaintainer(state);
+        }
+    }
+
+    @Test
+    void luceneOnlineIndexingTestTerribleRebalance() {
+        Index index = new Index("Simple$text_suffixes",
+                function(LuceneFunctionNames.LUCENE_TEXT, field("text")),
+                "terribleRebalance",
+                ImmutableMap.of(IndexOptions.TEXT_TOKENIZER_NAME_OPTION, AllSuffixesTextTokenizer.NAME));
+
+        boolean needMerge = populateDataSplitSegments(index, 4, 1);
+        assertTrue(needMerge);
+        try (OnlineIndexer indexBuilder = OnlineIndexer.newBuilder()
+                .setRecordStore(recordStore)
+                .setIndex(index)
+                .build()) {
+            indexBuilder.mergeIndex(); // retries under the hood until document count reaches minimum
+            assertTrue(recordStore.getIndexDeferredMaintenanceControl().getRepartitionDocumentCount() <= 2);
```

Comment:
It probably doesn't make sense with this extending `LuceneIndexMaintainer`, but something to think about if the deferred maintenance gets more complicated, or ever gets decoupled from lucene:
It may make sense to have it track the `MergeControl` controls, and then we can assert about exactly what it got called with during the index build. In order to do that.

Reply:
Do you mean by using `timer`?

Reply:
No, I meant tracking the `MergeControl`, i.e.:
assert that it got called with:
| repartition limit | merge limit | result |
| -- | -- | -- |
| 0 | 0 | failure on repartition |
| 8 | 0 | failure on repartition |
| 4 | 0 | failure on repartition |
| 2 | 0 | failure on merge |
| 0 | 8 | failure on merge |
| 0 | 4 | success |
...

Or something like that.